bindings/rust/wasmedge-sys/src/module.rs

```diff
@@ -1,33 +1,119 @@
 use super::wasmedge;
-use crate::raw_result::{decode_result, ErrReport};
+use crate::{
+    error::{check, WasmEdgeResult},
+    utils, Config,
+};
+use ::core::mem::MaybeUninit as MU;
 use std::ffi::CString;
+use std::path::Path;
 
 #[derive(Debug)]
 pub struct Module {
     pub(crate) ctx: *mut wasmedge::WasmEdge_ASTModuleContext,
+    pub(crate) registered: bool,
 }
-
 impl Drop for Module {
     fn drop(&mut self) {
-        unsafe { wasmedge::WasmEdge_ASTModuleDelete(self.ctx) };
+        if !self.registered && !self.ctx.is_null() {
+            unsafe { wasmedge::WasmEdge_ASTModuleDelete(self.ctx) };
+        }
     }
 }
-
 impl Module {
-    pub fn load_from_file(
+    pub fn load_from_file<P: AsRef<Path>>(config: &Config, path: P) -> WasmEdgeResult<Self> {
+        let loader_ctx = unsafe { wasmedge::WasmEdge_LoaderCreate(config.ctx) };
+        let mut ctx: *mut wasmedge::WasmEdge_ASTModuleContext = std::ptr::null_mut();
+
+        let path = utils::path_to_cstring(path.as_ref())?;
+
+        unsafe {
+            check(wasmedge::WasmEdge_LoaderParseFromFile(
+                loader_ctx,
+                &mut ctx as *mut _,
+                path.as_ptr(),
+            ))?;
+        }
+
+        Ok(Self {
+            ctx,
+            registered: false,
+        })
+    }
+
+    pub fn load_from_buffer(
         config: &crate::config::Config,
-        path: CString,
+        buffer: Vec<u8>,
     ) -> Result<Self, ErrReport> {
         let loader_ctx = unsafe { wasmedge::WasmEdge_LoaderCreate(config.ctx) };
         let mut ctx: *mut wasmedge::WasmEdge_ASTModuleContext = std::ptr::null_mut();
 
+        assert!(!buffer.is_empty(), "WasmEdge fail to load an empty buffer");
```

Comment:
nice! This assertion is guaranteed to be safety. 
This unsafe function should be marked with a document comment to indicate that it is safe, as is the usual practice. 
```rust
/// # Safety 
///  Avoid the problem of uninitialized memory due to the introduction of an empty buffer
```

Reply:
Thanks for your comments. This PR is replaced by #854, please add your comments there. Thanks again!